WordPress/gutenberg
# [RNMobile] Update Loading and Failed Screens for Unsupported Block Editor

Fixes the Android side of https://github.com/wordpress-mobile/gutenberg-mobile/issues/3425

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3573
`WordPress-Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/14762
`WordPress-iOS`: https://github.com/wordpress-mobile/WordPress-iOS/pull/16586

## Description

The `/sites/{site_id}/gutenberg` endpoint is [unreliable](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3425#issuecomment-840223555) but was being used to check if Gutenberg was available on a site's backend, and if so, disable the Unsupported Block Editor (UBE).

Since the majority of users will have Gutenberg enabled, it makes sense to remove usage of this unreliable API endpoint. That's done [in the Android PR](https://github.com/wordpress-mobile/WordPress-Android/pull/14762) in this PR chain.

To account for the removal, what we do in this PR is:

* Update the in-progress layout as the UBE loads, with a message warning users that if Gutenberg is not enabled on their site, the UBE will not load.
* Show a new failed screen if the UBE fails to load, e.g. because Gutenberg was not enabled on the site, including a message prompting users to check if Gutenberg is enabled on their site.

## How has this been tested?

To verify that the UBE now loads for Jetpack sites on an account created prior to December 2018:

* Sign in using an account created before Dec 2018.
* Locate a post with an unsupported block on an Atomic site.
* Open the post and tap the (?) to open the web editor (UBE).
* Ensure the UBE opens as expected and you can successfully edit the unsupported block.

The UBE test cases outlined [here](https://github.com/wordpress-mobile/test-cases/blob/dc78dcff1c7d01a3c6f10a023b053ad4566d3ba1/test-cases/gutenberg/unsupported-block-editing.md) were used as a guide for further testing:

```
- [ ] TC001 - User can edit unsupported blocks on Simple WP.com sites
- [ ] TC002 - User can discard edits to an unsupported blocks on Simple WP.com sites
- [ ] TC003 - Editing unsupported blocks is allowed on Gutenberg-enabled Atomic sites
- [ ] TC004 - Editing unsupported blocks is disallowed on Classic-enabled Atomic sites
- [ ] TC005 - Editing unsupported blocks is enabled on self-hosted sites accessed via Jetpack
- [ ] TC006 – Editing unsupported blocks is disallowed on self-hosted sites accessed via their own username and password
```

When going through each test case, verify that the following is true:

* The message warning users that Gutenberg is required is displayed when the UBE is loading.
* When testing `TC004`, the failed screen is displayed to users after the progress screen has been visible for 10 seconds. 

## Screenshots 

| UBE Progress Screen  | UBE Failed Screen |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/120470005-f48a0d00-c39a-11eb-985b-bcaf07d22d25.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/120469859-cf959a00-c39a-11eb-94a0-005366327972.png" width="100%"> |

## Types of changes

This PR introduces a non-breaking change, with the following summarising the main changes to the code:

* [The `foreground_view` displayed to users](https://github.com/WordPress/gutenberg/pull/32395/files#diff-952374fd3187d583fe1f5164645c83b74737d5192cec6dcfa907c2cf3a96cf9aR35) when the UBE is loading was fleshed out to include an image, a title, and a subtitle.
* A [new `showTroubleshootingInstructions()` function was introduced](https://github.com/WordPress/gutenberg/pull/32395/files#diff-5b6b9250e92e7fc0b8d9daa642b80ffc8906368041e02681113c68b9a9584831R82), which [fires if the UBE hasn't loaded after 10 seconds](https://github.com/WordPress/gutenberg/pull/32395/files#diff-5b6b9250e92e7fc0b8d9daa642b80ffc8906368041e02681113c68b9a9584831R72). This function updates the image, title, and subtitle displayed in the `foreground_view`.
* In line with the iOS change, the `img_illustration_cloud_off` image was taken [directly from the WordPress Android app](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/res/drawable/img_illustration_cloud_off_152dp.xml) and [is used in the `ube_failed.xml` file](https://github.com/WordPress/gutenberg/pull/32395/files#diff-572ea26ea944b0c83f2188d7e1c37cb74df3c43e5204ad17d5a3cf9088101ca3). The title and subtitle for the screens were also taken directly from the iOS change.

## Ideas Explored

As there isn't an equivalent to the loading animation used in the iOS side of this PR, the progress screen doesn't have an image. There is, however, an existing progress bar that's animated, to convey the idea that something is happening behind the scenes. 

Although I settled on no image on the progress screen, one could easily be added. Below are a couple of ones I experimented with on the Android side.

<details>
<summary>Images</summary>

| UBE Progress Screen  | UBE Failed Screen |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/120470231-3a46d580-c39b-11eb-9af8-d0d52552a96e.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/120470323-4d59a580-c39b-11eb-8377-4a83b3f3e0ac.png" width="100%"> |
</details>

## Checklist:
